Write secure HTTP API spike

tcharding opened the pull request:
As defined in the issue [below], we would like to consider securing the HTTP API exposed by cnd.  'securing' is an ambiguous term, this spike is limited in scope to encrypting the traffic between cnd and comit-i.

Fixes: https://github.com/comit-network/comit-rs/issues/477

tcharding committed:
Write secure HTTP API spike

Fixes: #477

bonomat reviewed (changes requested):
Thanks for the summary. 

I think you can go one step further and propose a solution to how we can secure cnd's API exactly. I'd say *not* doing this is not an option if we want to write *best practice software* as you state at the beginning.

tcharding commented:
> I'd say _not_ doing this is not an option if we want to write _best practice software_ as you state at the beginning.

Fair point, will do further research.

tcharding commented:
Ok, spike new and revised - I'm happy with my conclusions and that someone could now go ahead and implement TLS as recommended in this spike without having to do further research.  Please say if this view is incorrect - I'd really like to get this whole spike writing thing under my belt.

thomaseizinger reviewed (commented):
The problem with self-signed certs is that the browser will not trust them and display a giant warning to the user that what he tries to do is unsafe. IMO, this is not an acceptable solution.

If we want to use a self-signed certificate, I think we would need to install it as a CA on the computer (root privileges required) which seems to be bad.

This documentation here on letsencrypt.org goes into more detail for this problem: https://letsencrypt.org/docs/certificates-for-localhost/

Regarding the spike: I think the approach is good but don't think it will be possible to have a good solution to this if `comit-it` is out-of-scope :)

D4nte commented on 0013-secure-http-api.adoc:
```diff
@@ -0,0 +1,85 @@
+= Securing the `cnd` HTTP API
+
```
```suggestion
```
Must not have a new line for ascii doc to pick up the author line.

D4nte commented on 0013-secure-http-api.adoc:
```diff
@@ -0,0 +1,85 @@
+:toc:
+:revdate: 2019-08-13
+
+* Authors(s): {authors}
```
```suggestion
* Authors(s): {authors} +
* Date: {revdate} +
```
There is a '+' in the template, not sure what was the reason but pretty sure you need to keep it

thomaseizinger replied:
Is needed for correct rendering.

bonomat replied:
It's to get a new line after `* Authors(s): {authors}`
https://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#breaks

D4nte reviewed (commented):
@tcharding can you do a pass on the asciidoc format? Just check the file from GitHub and correct what does not render well (new line is needed before a list for example).

tcharding commented:
> @tcharding can you do a pass on the asciidoc format? Just check the file from GitHub and correct what does not render well (new line is needed before a list for example).

My bad, thanks for poking me.

tcharding commented:
> The problem with self-signed certs is that the browser will not trust them and display a giant warning to the user that what he tries to do is unsafe. IMO, this is not an acceptable solution.

Yes, I did not include that in the spike I assumed it was common knowledge and acceptable.

> If we want to use a self-signed certificate, I think we would need to install it as a CA on the computer (root privileges required) which seems to be bad.

This is worse IMO than just expecting the user to click through the warning - since our target are devs I feel it is safe to assume they know what doing that means and since they most likely started the cnd themselves this will not be a surprise.  Should this discussion go into the spike? 

> This documentation here on letsencrypt.org goes into more detail for this problem: [letsencrypt.org/docs/certificates-for-localhost](https://letsencrypt.org/docs/certificates-for-localhost/)
> 
> Regarding the spike: I think the approach is good but don't think it will be possible to have a good solution to this if `comit-it` is out-of-scope :)

Oh, does it read like `comit-i` is out of scope.  That was not my intention, I was hoping someone with front end knowledge would speak up.  Perhaps a better method of getting this information would be for me to actively ask :)

bonomat commented on 0013-secure-http-api.adoc:
In order to get some knowledge around comit-i, please pair up with someone and write it together.

tcharding replied:
> In order to get some knowledge around comit-i, please pair up with someone and write it together.

Ok

D4nte commented:
> This is worse IMO than just expecting the user to click through the warning - since our target are devs I feel it is safe to assume they know what doing that means and since they most likely started the cnd themselves this will not be a surprise. Should this discussion go into the spike?

Our target is devs that build apps used by users. In this case, the final user would still need to click through the warning if the app is the webapp which is not good.

_Also, please avoid starting discussions on the main thread for spikes, makes it hard to follow_

tcharding committed:
Correct author field

tcharding committed:
Fix list rendereing

tcharding committed:
Remove newline after heading

tcharding committed:
Simplify TLS implementation heading

tcharding committed:
Add section on usablity considerations re certs

tcharding commented:
> Also, please avoid starting discussions on the main thread for spikes, makes it hard to follow

Where should these discussions be please @D4nte?  On the original issue?

tcharding committed:
Fix rendering of authors section

thomaseizinger commented:
> Also, please avoid starting discussions on the main thread for spikes, makes it hard to follow

Probably my bad because I voiced the concern in the general review, with does not allow to create threads.

@tcharding 

Thanks for amending the spike with those concerns! How are we dealing with that problem now?

Didn't read it yet but maybe this HN thread has some insights: https://news.ycombinator.com/item?id=17358339

tcharding commented:
> Didn't read it yet but maybe this HN thread has some insights: news.ycombinator.com/item?id=17358339

Good link, thanks.  According to the blog post linked (indirectly) above; we may be able to get around using self signed certificates by using `127.0.0.1` instead of `localhost`.  Turns out modern browsers treat these addresses differently and the IP version is considered ["potentially trustworthy"](https://www.w3.org/TR/secure-contexts/#is-origin-trustworthy).

D4nte commented:
> > Also, please avoid starting discussions on the main thread for spikes, makes it hard to follow
> 
> Where should these discussions be please @D4nte? On the original issue?

In the "Files changed" tab, on the line concerned by the comment.

tcharding committed:
Add content on certificates

tcharding committed:
Clarify use cases

Clarify use cases for where cnd runs and how a user would connect to
`comit-i`.

Include section on let's encrypt suggestion to us IP 127.0.0.1 instead
of `localhost`.

D4nte commented on 0013-secure-http-api.adoc:
```suggestion
TLS requires a certificate (private/public keypair) to be present on the server i.e., cnd.
```

D4nte commented on 0013-secure-http-api.adoc:
```suggestion
1. Exposing the HTTP API over TLS (so clients can connect to cnd using TLS)
```

D4nte commented on 0013-secure-http-api.adoc:
```suggestion
2. Serving `comit-i` over TLS (so the browser can use `https://` addresses)
```

D4nte commented on 0013-secure-http-api.adoc:
I assume you just update the cnd REST API end point from `http` to `https`.

tcharding replied:
Unless I'm misunderstanding the architechture, I think it is more complicated than that.  `comit-i` consumes the REST API via typescript code, not via the web browser so `comit-i` has to do the TLS handshake.

D4nte commented on 0013-secure-http-api.adoc:
I don't understand why you would need that.

tcharding replied:
Because of my comment above, if it is correct.

D4nte commented on 0013-secure-http-api.adoc:
Why not:
- cnd generates certificates when starting up if not yet existing and if `webgui` is activated and `webgui.tls` (new option) is activated. This will be a CA, Public Cert and Private Cert
- User is told to install the freshly generated CA

That's it.

Feel free to at `lnd` as it does generate certificates for its gRPC API on the spot as described above (it does not generate a CA though, only a `tls.cert` and `tls.key`).

tcharding replied:
I think I get what you are saying but am confused by the terminology - I'm only just across all the TLS certificate terms as is.  Either way we still either make the user click through security warnings or require root to install the CA.

tcharding replied:
I do not think that `cnd` should create the certificate.  I think we should stick to the UNIX principle of doing one thing.  The user can generate the certificate.

thomaseizinger commented on 0013-secure-http-api.adoc:
```diff
@@ -52,34 +52,98 @@ The industry standard for encrypting traffic over a network is Transport Layer S
+* Leave generation of valid certificate up to the user of cnd but add documentation/hints on how to do this and what options are available (based on information in this spike).
+
+As for `comit-i` my findings were less positive.
+Recommend splitting `comit-i` up into separate parts for the front end and back end and re-writing the back end in Rust.
```
Is your idea that we can then serve `comit-i` on `http` again?
I am afraid, this does not really help us :(
One of the reasons why it would be useful to serve `comit-i` via HTTPS is to make integrations with a Ledger Nano S possible (the browser only allows you to talk to a USB device on HTTPS)

tcharding replied:
The idea is to serve `comit-i` using HTTPS.

thomaseizinger reviewed (approved):
I am starting to like this :)

It looks like our best option is to on-demand create a certificate for cnd if the user wants to have TLS for the API.

tcharding commented:
> It looks like our best option is to on-demand create a certificate for cnd if the user wants to have TLS for the API.

As commented above, I don't think we should be creating the certificate in `cnd`, I feel that we should leave it up to the user to do so.

luckysori committed:
Introduce final thoughts on this spike

thomaseizinger reviewed (approved):
Thanks for adding the final notes @luckysori !

FYI @tcharding, the final notes are the conclusion of our discussion this morning :)

Commit:
Merge branch 'master' into 477-secure-http-api
